ledger/accountdb.go

```diff
@@ -633,43 +1055,739 @@ func accountsAddNormalizedBalance(tx *sql.Tx, proto config.ConsensusParams) erro
-	if queryAddresses {
-		rows, err := tx.Query("SELECT address FROM accountbase where length(data) = 1 and data = x'80'") // empty AccountData is 0x80
+// accountsCreateResourceTable creates the resource table in the database.
+func accountsCreateResourceTable(ctx context.Context, tx *sql.Tx) error {
+	var exists bool
+	err := tx.QueryRowContext(ctx, "SELECT 1 FROM pragma_table_info('resources') WHERE name='addrid'").Scan(&exists)
+	if err == nil {
+		// Already exists.
+		return nil
+	}
+	if err != sql.ErrNoRows {
+		return err
+	}
+	for _, stmt := range createResourcesTable {
+		_, err = tx.ExecContext(ctx, stmt)
 		if err != nil {
-			return 0, nil, err
+			return err
 		}
-		defer rows.Close()
+	}
+	return nil
+}
 
-		for rows.Next() {
-			var addrbuf []byte
-			err = rows.Scan(&addrbuf)
-			if err != nil {
-				return 0, nil, err
-			}
-			var addr basics.Address
-			if len(addrbuf) != len(addr) {
-				err = fmt.Errorf("Account DB address length mismatch: %d != %d", len(addrbuf), len(addr))
-				return 0, nil, err
-			}
-			copy(addr[:], addrbuf)
-			addresses = append(addresses, addr)
-		}
+type baseOnlineAccountData struct {
+	_struct struct{} `codec:",omitempty,omitemptyarray"`
+
+	VoteID          crypto.OneTimeSignatureVerifier `codec:"A"`
+	SelectionID     crypto.VRFVerifier              `codec:"B"`
+	VoteFirstValid  basics.Round                    `codec:"C"`
+	VoteLastValid   basics.Round                    `codec:"D"`
+	VoteKeyDilution uint64                          `codec:"E"`
+	StateProofID    merklesignature.Verifier        `codec:"F"`
+}
+
+type baseAccountData struct {
+	_struct struct{} `codec:",omitempty,omitemptyarray"`
+
+	Status                     basics.Status     `codec:"a"`
+	MicroAlgos                 basics.MicroAlgos `codec:"b"`
+	RewardsBase                uint64            `codec:"c"`
+	RewardedMicroAlgos         basics.MicroAlgos `codec:"d"`
+	AuthAddr                   basics.Address    `codec:"e"`
+	TotalAppSchemaNumUint      uint64            `codec:"f"`
+	TotalAppSchemaNumByteSlice uint64            `codec:"g"`
+	TotalExtraAppPages         uint32            `codec:"h"`
+	TotalAssetParams           uint64            `codec:"i"`
+	TotalAssets                uint64            `codec:"j"`
+	TotalAppParams             uint64            `codec:"k"`
+	TotalAppLocalStates        uint64            `codec:"l"`
+
+	baseOnlineAccountData
+
+	// UpdateRound is the round that modified this account data last. Since we want all the nodes to have the exact same
+	// value for this field, we'll be setting the value of this field to zero *before* the EnableAccountDataResourceSeparation
+	// consensus parameter is being set. Once the above consensus takes place, this field would be populated with the
+	// correct round number.
+	UpdateRound uint64 `codec:"z"`
+}
+
+// IsEmpty return true if any of the fields other then the UpdateRound are non-zero.
+func (ba *baseAccountData) IsEmpty() bool {
+	return ba.Status == 0 &&
```

Comment:
why not use `return *ba == baseAccountData{}`

Reply:
it could also be used for the set and the get.
but as long as we don't "miss" something. it is good